File: app/src/main/java/com/kode/utair/presentation/mvp/presenters/CityWeatherForecastPresenter.java
package com.kode.utair.presentation.mvp.presenters;

import com.arellomobile.mvp.InjectViewState;
import com.arellomobile.mvp.MvpPresenter;
import com.kode.utair.di.ApplicationComponent;
import com.kode.utair.domain.exceptions.NoNetworkException;
import com.kode.utair.domain.interactors.WeatherForecastInteractor;
import com.kode.utair.presentation.mvp.views.CityWeatherForecastView;
import com.kode.utair.presentation.utils.DebugUtils;

import javax.inject.Inject;

@InjectViewState
public class CityWeatherForecastPresenter extends MvpPresenter<CityWeatherForecastView> {

    @Inject
    WeatherForecastInteractor weatherForecastInteractor;

    private String cityName;

    public CityWeatherForecastPresenter(ApplicationComponent applicationComponent, String cityName) {
        this.cityName = cityName;

        applicationComponent.inject(this);
        initUi();
    }

    private void initUi() {
        showForecast();
    }

    private void showForecast() {
        weatherForecastInteractor.getWeatherForecastForCity(cityName)
                .subscribe(weatherForecastEntity -> getViewState().showForecast(weatherForecastEntity.getDailyForecasts()),
                        this::handleError);
    }

    private void handleError(Throwable throwable) {
        if (throwable instanceof NoNetworkException) {
            getViewState().showNoNetworkMessage();
        } else {
            DebugUtils.showDebugErrorMessage(throwable);
        }
    }

}

File: app/src/main/java/com/kode/utair/presentation/mvp/presenters/CityWeatherForecastPresenter.kt
package com.kode.utair.presentation.mvp.presenters

import android.util.Log
import com.arellomobile.mvp.InjectViewState
import com.arellomobile.mvp.MvpPresenter
import com.kode.utair.domain.exceptions.NoNetworkException
import com.kode.utair.domain.interactors.WeatherForecastInteractor
import com.kode.utair.domain.models.WeatherForecastEntity
import com.kode.utair.presentation.mvp.views.CityWeatherForecastView
import com.kode.utair.presentation.utils.DebugUtils

@InjectViewState
class CityWeatherForecastPresenter(
        private val weatherForecastInteractor: WeatherForecastInteractor,
        private val cityName: String
) : MvpPresenter<CityWeatherForecastView>() {

    override fun onFirstViewAttach() {
        super.onFirstViewAttach()

        showForecast()
    }

    private fun showForecast() {
        weatherForecastInteractor.getWeatherForecastForCity(cityName)
                .subscribe(
                        { weatherForecastEntity: WeatherForecastEntity ->
                            viewState.showForecast(weatherForecastEntity.dailyForecasts)
                        })
                { throwable: Throwable -> handleError(throwable) }
    }

    private fun handleError(throwable: Throwable) {
        Log.e("UTair", "showForecast", throwable)
        if (throwable is NoNetworkException) {
            viewState.showNoNetworkMessage()
        } else {
            DebugUtils.showDebugErrorMessage(throwable)
        }
    }

}